### PR TITLE
Add newlines in type assignment errors (#45896)

### LINKED ISSUE #45896: Use the width of a type to add newlines in type assignment error messages
Today we ship a one-type-fits-all style for printing type is not assignable to type messages. We'd like to improve this in a pretty simple manner: by occasionally adding newlines. For example with this arbitrary comparison:

```ts
let a = { b: { c: { e: { f: 123 } } } };
let b = { b: { c: { e: { f: "123" } } } };
a = b;
```

Looks like this today:

```sh
src/vendor/ata/index.ts(12,1): error TS2322: Type '{ b: { c: { e: { f: string; }; }; }; }' is not assignable to type '{ b: { c: { e: { f: number; }; }; }; }'.
The types of 'b.c.e.f' are incompatible between these types.
    Type 'string' is not assignable to type 'number'.
```

I propose that because both of the printed types are longer than 20-30 chars, then we switch to:

```sh
src/vendor/ata/index.ts(12,1): error TS2322: Type: 
{ b: { c: { e: { f: string; }; }; }; }

is not assignable to type:
{ b: { c: { e: { f: number; }; }; }; }

The types of 'b.c.e.f' are incompatible between these types.
    Type 'string' is not assignable to type 'number'.
```

### Key details

 - Strip the quotes around the type
 - Add colons and newlines if the type is over a certain size. Let's call it 30 characters for now, and make it easy to change and we can experiment at the final stages of review.

#### No changes

```ts
let a = 123
let b = "abc"
a = b
```

- Before: `src/index.ts(21,1): error TS2322: Type 'string' is not assignable to type 'number'.`
- After: `src/index.ts(21,1): error TS2322: Type 'string' is not assignable to type 'number'.`

No change! (Though a part of me does really want to drop the quotes for primitives)

```ts
window = {}
```

```
Type '{}' is not assignable to type 'Window & typeof globalThis'.
  Type '{}' is missing the following properties from type 'Window': clientInformation, closed, customElements, devicePixelRatio, and 189 more. (2322)
```

No change! `{}` and `Window & typeof globalThis ` are too small.

#### Changing one

```ts
window = addEventListener
```

```
src/index.ts(1,1): error TS2322: Type '{ <K extends keyof WindowEventMap>(type: K, listener: (this: Window, ev: WindowEventMap[K]) => any, options?: boolean | AddEventListenerOptions | undefined): void; (type: string, listener: EventListenerOrEventListenerObject, options?: boolean | ... 1 more ... | undefined): void; }' is not assignable to type 'Window & typeof globalThis'.
   Type '{ <K extends keyof WindowEventMap>(type: K, listener: (this: Window, ev: WindowEventMap[K]) => any, options?: boolean | AddEventListenerOptions | undefined): void; (type: string, listener: EventListenerOrEventListenerObject, options?: boolean | ... 1 more ... | undefined): void; }' is missing the following properties from type 'Window': HTMLDocument, closed, customElements, devicePixelRatio, and 187 more.
```

```
src/vendor/ata/index.ts(12,1): error TS2322: Type: 
{ <K extends keyof WindowEventMap>(type: K, listener: (this: Window, ev: WindowEventMap[K]) => any, options?: boolean | AddEventListenerOptions | undefined): void; (type: string, listener: EventListenerOrEventListenerObject, options?: boolean | ... 1 more ... | undefined): void; }

is not assignable to type:
Window & typeof globalThis

  Type '{ <K extends keyof WindowEventMap>(type: K, listener: (this: Window, ev: WindowEventMap[K]) => any, options?: boolean | AddEventListenerOptions | undefined): void; (type: string, listener: EventListenerOrEventListenerObject, options?: boolean | ... 1 more ... | undefined): void; }' is missing the following properties from type 'Window': HTMLDocument, closed, customElements, devicePixelRatio, and 187 more.
```

#### Changing both

```
Animation = AnalyserNode
```

```
index.ts(12,1) Type '{ new (context: BaseAudioContext, options?: AnalyserOptions | undefined): AnalyserNode; prototype: AnalyserNode; }' is not assignable to type '{ new (effect?: AnimationEffect | null | undefined, timeline?: AnimationTimeline | null | undefined): Animation; prototype: Animation; }'.
  Types of property 'prototype' are incompatible.
    Type 'AnalyserNode' is missing the following properties from type 'Animation': currentTime, effect, finished, id, and 18 more.
```

```
index.ts(12,1) Type:
{ new (context: BaseAudioContext, options?: AnalyserOptions | undefined): AnalyserNode; prototype: AnalyserNode; }

is not assignable to type:
{ new (effect?: AnimationEffect | null | undefined, timeline?: AnimationTimeline | null | undefined): Animation; prototype: Animation; }

  Types of property 'prototype' are incompatible.
    Type 'AnalyserNode' is missing the following properties from type 'Animation': currentTime, effect, finished, id, and 18 more.
```

### PR DESCRIPTION
**Pull request by cyntler**
Hey!
I committed some changes related to the issue: https://github.com/microsoft/TypeScript/issues/45896.
That's my first PR to the TypeScript repo!
I hope it's okay and I am waiting for code review.

Fixes #45896

### CODE REVIEW

**Comment by cyntler** [comment]
@orta Conflict resolved.

**Comment by orta** [comment]
Thanks! I'll make a build and give this a run through

**Comment by orta** [comment]
@typescript-bot pack this

**Comment by cyntler** [comment]
@orta Hey! I resolved conflict again. Have you looked at it?

**Comment by orta** [comment]
I'm a little wary of the amount of newlines in message this can cause after playing with it, so I've made it check the length of each before printing and only modifying a part if it's needed - will give it a run in the playground, but I might be able to get this into 4.5 if I'm lucky

@typescript-bot pack this

File: tests/baselines/reference/classPropertyErrorOnNameOnly.errors.txt [hunk 1]
```diff
@@ -2,7 +2,10 @@ tests/cases/compiler/classPropertyErrorOnNameOnly.ts(7,3): error TS2322: Type '(
   Type 'string | undefined' is not assignable to type 'string'.
     Type 'undefined' is not assignable to type 'string'.
 tests/cases/compiler/classPropertyErrorOnNameOnly.ts(24,7): error TS2322: Type '(val: Values) => "1" | "2" | "3" | "4" | "5" | undefined' is not assignable to type 'FuncType'.
-  Type 'string | undefined' is not assignable to type 'string'.
+  Type:
```
Comment (sandersn) [inline tests/baselines/reference/classPropertyErrorOnNameOnly.errors.txt:5]: reformatting is still happening in elaborations

**Review by sandersn — CHANGES_REQUESTED** [review]

**Comment by orta** [comment]
Heh, switching it to `pretty` only means no baselines are updated, will generate a new one

**Comment by orta** [comment]
@typescript-bot pack this

**Comment by orta** [comment]
@typescript-bot user test this

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

**Comment by RyanCavanaugh** [comment]
🧹 Housekeeping: Closing this PR since it's been a while since changes were requested and we haven't seen any activity recently. Please open a new PR if you'd like to continue with this. Thanks!

**Pull request closed without merging**